File: Game2D/Project/Boss.java
import greenfoot.*;
import java.util.*;

public class Boss extends Actor {
    private int floatOffset = 0;
    private int direction = 1;
    private double koef=1;
    private int maxHealth = (int)(1000 * koef);
    private int health = maxHealth;
    private long lastAttackTime = 0;
    private int difficultyLevel = 1;
    private long startTime;
    private long lastDifficultyCheck;
    private Queue<Integer> attackQueue = new LinkedList<>();
    private double attackSpeedMultiplier = 1.0;
    private int baseAttackCooldown = 5000;
    private GreenfootImage healthBar;
    private int healthBarWidth = 400;
    private int healthBarHeight = 25;
    private HealthBar bar;
    private Hero hero;
    
    public Boss(double coef, Hero h) {
        startTime = System.currentTimeMillis();
        lastDifficultyCheck = startTime;
        initializeAttackQueue();
        this.koef = coef;
        maxHealth = (int)(maxHealth * koef);
        health = maxHealth;
        this.attackSpeedMultiplier = (double)(1.0 + coef);
        this.baseAttackCooldown = (int)(baseAttackCooldown / (1+coef));
        this.hero = h;
        
        healthBar = new GreenfootImage(healthBarWidth, healthBarHeight);
        updateHealthBar();
        
    }

    public void act() {
        if (getWorld() == null || hero==null) return;

        if (bar == null) {
            bar = new HealthBar(healthBar);
            getWorld().addObject(bar, getX(), getY() - 80);
        }

        updateDifficulty();
        updateAttackSpeed();
        floatUpDown();

        long currentTime = System.currentTimeMillis();
        int currentCooldown = (int)(baseAttackCooldown / attackSpeedMultiplier);

        if (currentTime - lastAttackTime > currentCooldown) {
            startNextAttack();
            lastAttackTime = currentTime;
        }

        
        if (bar != null) {
            bar.setLocation(getX(), getY() - 80);
        }
    }
    
    private void updateHealthBar() {
        healthBar.clear();
        healthBar.setColor(Color.BLACK);
        healthBar.fill();
        healthBar.setColor(Color.GRAY);
        healthBar.fillRect(1, 1, healthBarWidth - 2, healthBarHeight - 2);

        int currentWidth = (int)((double)health / maxHealth * (healthBarWidth - 4));
        healthBar.setColor(Color.RED);
        healthBar.fillRect(2, 2, currentWidth, healthBarHeight - 4);

        if (bar != null) {
            bar.setImage(healthBar);
        }
    }
    
    private void updateAttackSpeed() {
        long elapsed = (System.currentTimeMillis() - startTime) / 1000;
        attackSpeedMultiplier = 1.0 + (elapsed / 15.0) * 0.5;
        attackSpeedMultiplier = Math.min(attackSpeedMultiplier, 3.0);
    }
    
    private void initializeAttackQueue() {
        attackQueue.clear();
        attackQueue.add(0); // Вертикальная
        attackQueue.add(1); // Горизонтальная
        attackQueue.add(2); // Диагональная
        attackQueue.add(3); // Комбо
        attackQueue.add(4); // Пулемет
    }
    
    private void startNextAttack() {
        if (attackQueue.isEmpty()) {
            initializeAttackQueue();
        }
        
        int nextAttack = attackQueue.poll();
        
        switch (nextAttack) {
            case 0: attackVertical(difficultyLevel); break;
            case 1: attackHorizontal(difficultyLevel); break;
            case 2: attackDiagonal(difficultyLevel); break;
            case 3: attackSpecialCombo(); break;
            case 4: attackBulletSpam(); break;
        }
        
        attackQueue.add(nextAttack);
    }
    
    private void updateDifficulty() {
        long currentTime = System.currentTimeMillis();
        long elapsed = (currentTime - startTime) / 1000;
        
        if ((currentTime - lastDifficultyCheck) > 10000) { 
            lastDifficultyCheck = currentTime;
            
            if (elapsed > 60) {
                difficultyLevel = 3;
            } else if (elapsed > 30) {
                difficultyLevel = 2;
            } else {
                difficultyLevel = 1;
            }
        }
    }

    private void floatUpDown() {
        setLocation(getX(), getY() + direction);
        floatOffset += direction;

        if (Math.abs(floatOffset) >= 10) {
            direction *= -1;
        }
    }

    private void attackVertical(int level) {
    World world = getWorld();
    if (world == null) return;

    int worldWidth = world.getWidth();
    int worldHeight = world.getHeight();

  
    int[][] spawnPositions = {
        {worldWidth / 5, worldHeight - 1, -1},  // слева, сверху вниз (направление -1)
        {2 * worldWidth / 5, 0, 1},            // левее центра, снизу вверх (направление +1)
        {3 * worldWidth / 5, worldHeight - 1, -1}, // правее центра, сверху вниз (направление -1)
        {4 * worldWidth / 5, 0, 1}              // справа, снизу вверх (направление +1)
    };

    List<int[]> spawnList = Arrays.asList(spawnPositions);
    Collections.shuffle(spawnList); // перемешиваем позиции

    int handsToSpawn = 2 + new java.util.Random().nextInt(2); 
    java.util.Random rand = new java.util.Random();

    for (int i = 0; i < handsToSpawn; i++) {
        int[] pos = spawnList.get(i);
        
        // Случайный выбор между Boss_Hand1 и Boss_Hand2
        Actor hand;
        if (rand.nextBoolean()) {
            hand = new Boss_Hand1();
            
        } else {
            hand = new Boss_Hand2();
        }
        
       
        if (hand instanceof Boss_Hand1) {
            Boss_Hand1 hand1 = (Boss_Hand1) hand;
            hand1.set_damage(100 + (int)(koef * 20));
            hand1.setMoveType(0); // диагональное движение
            hand1.setDirection(pos[2]);
            hand1.setSpeed(1 + level);
            // Поворот руки: -45° если вверх, +45° если вниз
            hand1.setRotation(pos[2] == 1 ? -45 : 45);
        } else if (hand instanceof Boss_Hand2) {
            Boss_Hand2 hand2 = (Boss_Hand2) hand;
            hand2.set_damage(100 + (int)(koef * 20));
            hand2.setMoveType(0); 
            hand2.setDirection(pos[2]);
            hand2.setSpeed(1 + level);
            // Поворот руки: -45° если вверх, +45° если вниз
            hand2.setRotation(pos[2] == 1 ? -45 : 45);
        }

        world.addObject(hand, pos[0], pos[1]);
    }
}

    private void attackHorizontal(int level) {
       World world = getWorld();
        if (world == null) return;

        int[] ySlots = {
            getWorld().getHeight() / 4,
            getWorld().getHeight() / 2,
            3 * getWorld().getHeight() / 4
        };

        List<Integer> indices = Arrays.asList(0, 1, 2);
        Collections.shuffle(indices);

        for (int i = 0; i < 2; i++) {
            int yIndex = indices.get(i);
            int y = ySlots[yIndex];

            if (Greenfoot.getRandomNumber(2) == 0) {
                Boss_Hand1 hand = new Boss_Hand1();
                hand.set_damage(100 + (int)(koef * 20));
                hand.setDirection(-1);
                hand.setMoveType(1);
                hand.setSpeed(1 + level);
                hand.setRotation(45);
                world.addObject(hand, world.getWidth() - 1, y);
            } else {
                Boss_Hand2 hand = new Boss_Hand2();
                hand.set_damage(100 + (int)(koef * 20));
                hand.setDirection(1);
                hand.setMoveType(1);
                hand.setSpeed(1 + level);
                hand.setRotation(-45);
                world.addObject(hand, 0, y);
            }
        } 
    }

    private void attackDiagonal(int level) {
        World world = getWorld();
        if (world == null) return;

        int width = world.getWidth();
        int height = world.getHeight();

        for (int i = 0; i < level; i++) {
            Boss_Hand2 hand2 = new Boss_Hand2();
            hand2.set_damage(100 + (int)(koef * 20));
            hand2.setDirection(1);
            hand2.setMoveType(2);
            world.addObject(hand2, i * 50, i * 50);

            Boss_Hand1 hand1 = new Boss_Hand1();
            hand1.set_damage(100 + (int)(koef * 20));
            hand1.setDirection(-1);
            hand1.setMoveType(2);
            hand1.setRotation(90);
            world.addObject(hand1, width - 1 - i * 50, height - 1 - i * 50);

            Boss_Hand1 hand3 = new Boss_Hand1();
            hand3.set_damage(100 + (int)(koef * 20));
            hand3.setDirection(-1);
            hand3.setMoveType(3);
            hand3.setRotation(180);
            world.addObject(hand3, 0 + i * 50, height - 1 - i * 50);

            Boss_Hand2 hand4 = new Boss_Hand2();
            hand4.set_damage(100 + (int)(koef * 20));
            hand4.setDirection(1);
            hand4.setMoveType(3);
            hand4.setRotation(90);
            world.addObject(hand4, width - 1 - i * 50, 0 + i * 50);
        }
    }

    private void attackSpecialCombo() {
    World world = getWorld();
    if (world == null) return;

    for (int i = 0; i < difficultyLevel; i++) {
       
        int offsetX = i * 50;
        int offsetY = 20 + i * 10;

        // Левая рука (ComboHand1)
        Boss_ComboHand1 hand = new Boss_ComboHand1();
        hand.set_damage(200 + (int)(koef * 20));
        hand.setAttackOrder(i);
        hand.setDifficultyLevel(difficultyLevel);
        world.addObject(
            hand,
            getX() + 150 + offsetX,
            getY() + offsetY
        );

        // Правая рука (ComboHand2)
        Boss_ComboHand2 rightHand = new Boss_ComboHand2();
        hand.set_damage(200 + (int)(koef * 20));
        rightHand.setAttackOrder(i);
        rightHand.setDifficultyLevel(difficultyLevel);
        world.addObject(
            rightHand,
            getX() - 150 - offsetX,
            getY() + offsetY
        );
    }
}

    private void attackBulletSpam() {
    World world = getWorld();
    if (world == null) return;

    int pattern = Greenfoot.getRandomNumber(3) + 1; 
    int offset = 400;

    switch (pattern) {
        case 1:
            // Обычное расположение по горизонтали
            ShooterHand1 hand1 = new ShooterHand1();
            hand1.set_coef(koef);
            
            world.addObject(hand1, getX() + offset, getY());

            ShooterHand2 hand2 = new ShooterHand2();
             hand2.set_coef(koef);
            world.addObject(hand2, getX() - offset, getY());
            break;

        case 2:
            // Зеркальное расположение, под диагональю, с поворотом на 90
            ShooterHand1 diagHand1 = new ShooterHand1();
            diagHand1.set_coef(koef);
            diagHand1.setRotation(180);
            world.addObject(diagHand1, getX() - offset, getY() + offset);

            ShooterHand2 diagHand2 = new ShooterHand2();
            diagHand2.set_coef(koef);
            diagHand2.setRotation(180);
            world.addObject(diagHand2, getX() + offset, getY() + offset);
            break;

        case 3:
            // Квадрат из 4-х рук
           ShooterHand2 topLeft = new ShooterHand2();
             topLeft.set_coef(koef);
        world.addObject(topLeft, getX() - offset, getY());

        ShooterHand1 topRight = new ShooterHand1();
         topRight.set_coef(koef);
        world.addObject(topRight, getX() + offset, getY());

        ShooterHand1 bottomLeft = new ShooterHand1();
        bottomLeft.set_coef(koef);
        bottomLeft.setRotation(180);
        world.addObject(bottomLeft, getX() - offset, getY() + offset + 50);

        ShooterHand2 bottomRight = new ShooterHand2();
        bottomRight.set_coef(koef);
        bottomRight.setRotation(180);
        world.addObject(bottomRight, getX() + offset, getY() + offset + 50);
            break;
    }
    }

    public void takeDamage(int amount) {
        health -= amount * difficultyLevel;
        if (health < 0) health = 0;
        updateHealthBar();

        if (health <= 0 && getWorld() != null) {
            getWorld().addObject(new Win() , getWorld().getWidth()/2,getWorld().getHeight()/2);
            getWorld().removeObject(this);
            Greenfoot.stop();
        }
    }
    
    public int getDifficultyLevel() {
        return difficultyLevel;
    }
    public double getAttackSpeedMultiplier() {
        return attackSpeedMultiplier;
    }
    
    class HealthBar extends Actor {
        public HealthBar(GreenfootImage image) {
            setImage(image);
        }

        public void act() {
 
        }
    }
}

File: Game2D/Engine/Boss.cs
using System.Windows.Controls;
using System.Windows.Media.Imaging;
using System.Windows.Shapes;
using System.Windows.Media;
using System.Collections.Generic;
using System;

namespace Game2D.Engine
{
    public class Boss : GameObject
    {
        private Hero _hero;
        private double koef = 1;
        private int maxHealth = 1500;
        private int health;
        private int attackCooldown = 180;
        private int attackTimer = 0;
        private Image bossSprite;
        private Rectangle hpBar;
        private double hpBarWidth = 400;
        private double hpBarHeight = 24;
        public bool IsDead => health <= 0;
        
        private int difficultyLevel = 1;
        private double attackSpeedMultiplier = 1.0;
        private Queue<string> attackQueue = new Queue<string>();
        private int currentAttackPhase = 0;
        private int phaseTimer = 0;
        private int phaseDuration = 900;
        private double floatOffset = 0;
        private double floatSpeed = 0.05;
        
        private readonly string[] attackTypes = { "vertical", "horizontal", "diagonal", "circle", "bulletSpam", "specialCombo" };

        public Boss(Hero hero, double x, double y, double coef = 1)
        {
            X = x;
            Y = y;
            _hero = hero;
            koef = coef;
            maxHealth = (int)(1500 * koef);
            health = maxHealth;
            bossSprite = new Image
            {
                Source = new BitmapImage(new Uri("pack://application:,,,/Project/images/boss.png")),
                Width = 128,
                Height = 128
            };
            Sprite = bossSprite;
            hpBar = new Rectangle
            {
                Width = hpBarWidth,
                Height = hpBarHeight,
                Fill = Brushes.Red,
                Stroke = Brushes.Black,
                StrokeThickness = 2
            };

            InitializeAttackQueue();
        }

        private void InitializeAttackQueue()
        {
            attackQueue.Clear();
            attackQueue.Enqueue("vertical");
            attackQueue.Enqueue("horizontal");
            attackQueue.Enqueue("diagonal");
            attackQueue.Enqueue("circle");
            
            if (difficultyLevel >= 2)
            {
                attackQueue.Enqueue("bulletSpam");
            }
            if (difficultyLevel >= 3)
            {
                attackQueue.Enqueue("specialCombo");
            }
        }

        public override void Update()
        {
            if (IsDead) return;
            
            UpdateDifficulty();
            
            FloatUpDown();
            
            phaseTimer++;
            if (phaseTimer >= phaseDuration)
            {
                phaseTimer = 0;
                currentAttackPhase++;
                StartNextAttack();
            }
            
            attackTimer++;
            if (attackTimer >= attackCooldown * attackSpeedMultiplier)
            {
                attackTimer = 0;
                PerformCurrentAttack();
            }
        }

        private void UpdateDifficulty()
        {
            double healthPercent = (double)health / maxHealth;
            if (healthPercent <= 0.2 && difficultyLevel < 4)
            {
                difficultyLevel = 4;
                attackSpeedMultiplier = 1.5;
            }
            else if (healthPercent <= 0.4 && difficultyLevel < 3)
            {
                difficultyLevel = 3;
                attackSpeedMultiplier = 1.3;
            }
            else if (healthPercent <= 0.6 && difficultyLevel < 2)
            {
                difficultyLevel = 2;
                attackSpeedMultiplier = 1.1;
            }
        }

        private void FloatUpDown()
        {
            floatOffset += floatSpeed;
            double floatY = Y + Math.Sin(floatOffset) * 5;
            Y = floatY;
        }

        private void StartNextAttack()
        {
            if (attackQueue.Count > 0)
            {
                string nextAttack = attackQueue.Dequeue();
                attackQueue.Enqueue(nextAttack);
                
                
                switch (nextAttack)
                {
                    case "vertical":
                        AttackVertical(difficultyLevel);
                        break;
                    case "horizontal":
                        AttackHorizontal(difficultyLevel);
                        break;
                    case "diagonal":
                        AttackDiagonal(difficultyLevel);
                        break;
                    case "circle":
                        AttackCircle(difficultyLevel);
                        break;
                    case "bulletSpam":
                        AttackBulletSpam();
                        break;
                    case "specialCombo":
                        AttackSpecialCombo();
                        break;
                }
            }
        }

        private void PerformCurrentAttack()
        {
            Attack();
        }

        private void AttackVertical(int level)
        {
            var world = MainWindow.CurrentGameWorld;
            if (world == null) return;

            int bulletCount = Math.Min(level + 1, 4);
            double spread = 0.2;
            
            for (int i = 0; i < bulletCount; i++)
            {
                double angle = -Math.PI/2 + (i - bulletCount/2.0) * spread;
                var rocket = new Rocket(world._canvas, X + Sprite.Width/2, Y + Sprite.Height/2, angle);
                world.AddObject(rocket);
            }
        }

        private void AttackHorizontal(int level)
        {
            var world = MainWindow.CurrentGameWorld;
            if (world == null) return;

            int bulletCount = Math.Min(level + 2, 5);
            double spread = 0.3;
            
            for (int i = 0; i < bulletCount; i++)
            {
                double angle = (i - bulletCount/2.0) * spread;
                var rocket = new Rocket(world._canvas, X + Sprite.Width/2, Y + Sprite.Height/2, angle);
                world.AddObject(rocket);
            }
        }

        private void AttackDiagonal(int level)
        {
            var world = MainWindow.CurrentGameWorld;
            if (world == null) return;
            double[] angles = { Math.PI/4, 3*Math.PI/4, 5*Math.PI/4, 7*Math.PI/4 };
            
            foreach (double angle in angles)
            {
                var rocket = new Rocket(world._canvas, X + Sprite.Width/2, Y + Sprite.Height/2, angle);
                world.AddObject(rocket);
            }
        }

        private void AttackCircle(int level)
        {
            var world = MainWindow.CurrentGameWorld;
            if (world == null) return;
            
            int bulletCount = 6 + level;
            double angleStep = 2 * Math.PI / bulletCount;
            
            for (int i = 0; i < bulletCount; i++)
            {
                double angle = i * angleStep;
                var rocket = new Rocket(world._canvas, X + Sprite.Width/2, Y + Sprite.Height/2, angle);
                world.AddObject(rocket);
            }
        }

        private void AttackBulletSpam()
        {
            var world = MainWindow.CurrentGameWorld;
            if (world == null) return;

            int pattern = new Random().Next(3) + 1;
            int offset = 400;

            switch (pattern)
            {
                case 1:
                    var hand1 = new ShooterHand1(X + offset, Y, _hero, koef);
                    world.AddGameObject(hand1);

                    var hand2 = new ShooterHand2(X - offset, Y, _hero, koef);
                    world.AddGameObject(hand2);
                    break;

                case 2:
                    var diagHand1 = new ShooterHand1(X - offset, Y + offset, _hero, koef);
                    world.AddGameObject(diagHand1);

                    var diagHand2 = new ShooterHand2(X + offset, Y + offset, _hero, koef);
                    world.AddGameObject(diagHand2);
                    break;

                case 3:
                    var topLeft = new ShooterHand2(X - offset, Y, _hero, koef);
                    world.AddGameObject(topLeft);

                    var topRight = new ShooterHand1(X + offset, Y, _hero, koef);
                    world.AddGameObject(topRight);

                    var bottomLeft = new ShooterHand1(X - offset, Y + offset + 50, _hero, koef);
                    world.AddGameObject(bottomLeft);

                    var bottomRight = new ShooterHand2(X + offset, Y + offset + 50, _hero, koef);
                    world.AddGameObject(bottomRight);
                    break;
            }
        }

        private void AttackSpecialCombo()
        {
            var world = MainWindow.CurrentGameWorld;
            if (world == null) return;
            
            double bossX = X + Sprite.Width/2;
            double bossY = Y + Sprite.Height/2;
            
            var comboHand1 = new Boss_ComboHand1(bossX - 100, bossY + 50, _hero, 1);
            var comboHand2 = new Boss_ComboHand2(bossX + 100, bossY + 50, _hero, 2);
            
            world.AddObject(comboHand1);
            world.AddObject(comboHand2);
        }

        private void Attack()
        {
            double dx = _hero.X + _hero.Sprite.Width / 2 - (X + Sprite.Width / 2);
            double dy = _hero.Y + _hero.Sprite.Height / 2 - (Y + Sprite.Height / 2);
            double angle = Math.Atan2(dy, dx);
            var world = MainWindow.CurrentGameWorld;
            if (world != null)
            {
                var canvas = world._canvas;
                var rocket = new Rocket(canvas, X + Sprite.Width / 2, Y + Sprite.Height / 2, angle);
                world.AddObject(rocket);
            }
        }

        public void TakeDamage(int dmg)
        {
            health -= dmg;
            if (health < 0) health = 0;
        }

        public int GetDifficultyLevel() => difficultyLevel;
        public double GetAttackSpeedMultiplier() => attackSpeedMultiplier;

        public override void Draw(Canvas canvas)
        {
            base.Draw(canvas);
            if (hpBar != null)
            {
                double percent = (double)health / maxHealth;
                hpBar.Width = hpBarWidth * percent;
                Canvas.SetLeft(hpBar, (canvas.ActualWidth - hpBarWidth) / 2);
                Canvas.SetTop(hpBar, 30);
                if (!canvas.Children.Contains(hpBar))
                    canvas.Children.Add(hpBar);
            }
        }
    }
} 